src/libstd/io/buffered.rs

```diff
@@ -451,7 +453,14 @@ impl<W: Write> BufWriter<W> {
     pub fn into_inner(mut self) -> Result<W, IntoInnerError<BufWriter<W>>> {
         match self.flush_buf() {
             Err(e) => Err(IntoInnerError(self, e)),
-            Ok(()) => Ok(self.inner.take().unwrap())
+            Ok(()) => unsafe {
+                // use `ptr::read` to extract the inner writer and circumvent
+                // our `Drop` implementation
+                mem::replace(&mut self.buf, Vec::new());
```

Comment:
Ah, now I understand why you prefer `mem::replace`. Both are fine with me I guess, although I think a short comment that `buf` is being dropped here probably wouldn't go amiss.